### ISSUE TITLE
Intellisense fails to infer result type of lambas beyond "__result_type"

### ISSUE THREAD

**Issue by gudspelr** [microsoft/vscode-cpptools#2774]
```c++
auto vsCodeFail() {
  auto step1 = []() { return make_tuple(1, string()); };
  using resultType = decltype(step1());
  auto step2 = [](resultType const& me) {
    // shows "__result_type' when hovering over resultType instead of std::tuple<int, std::string>
    auto const& [theInt, theString] = me;
    // shows red squiggly under [ and
    // shows <error-type> theInt, and <error-type>theString> when hovering
  };
}
```
NB: Visual Studio intellisense manages to at least infer that the result type is a tuple, but, it also fails to infer the types in the structured binding

**Comment by sean-mcmanus** [comment]
It's not reproing for me. What OS and compilerPath, intelliSenseMode, etc. are you using in c_cpp_properties.json. I assume you're on 0.20.1 of the extension?

![image](https://user-images.githubusercontent.com/19859882/48041860-00a28500-e134-11e8-864b-5905f5ac82bb.png)

**Comment by gudspelr** [comment]
I didn't realize the c_cpp_properties would have an impact.  The c_cpp_properties is for linux, but I was opening it in Windows.  Opening the same folder under linux with the same :
```
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64",
```
 yields effectively the same results, except I have `<type>` instead of `<error-type>` for the structured binding, and no squiqqly under the "[".  

However, deleting the c_cpp_properties results in it showing the correct type for lambda return type under windows, but not under linux.

In all cases, structured binding type inferencing fails

**Comment by gudspelr** [comment]
Interestingly, in linux, hovering over the word `decltype` shows the correct return type.

**Comment by sean-mcmanus** [comment]
Okay, I'm able to repro all the issues when using gcc as the compiler on Windows -- it says "cannot instantiate `std::tuple_element` for `<0, "const resultType">`. It doesn't repro when using cl.exe/msvc-x64 mode (and the red squiggle doesn't appear with mingw or clang+Linux).

**Comment by sean-mcmanus** [comment]
FYI, the error squiggle is fixed with [0.21.0-insiders2](https://github.com/Microsoft/vscode-cpptools/releases), but the hover type info still says __result_type, `<type>`, etc.. I filed a VS bug for the "type" issue at https://developercommunity.visualstudio.com/content/problem/416902/c-intellisense-shows-instead-of-the-actual-type-wi.html . The __result_type issue might be related or it could be different, I suspect related to the library implementation and typedefs involved.

**Comment by sean-mcmanus** [comment]
The __result_type issue hover bug should be fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/0.26.3-insiders .
